flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -1181,137 +1184,284 @@ struct EmboxOpConversion : public FIROpConversion<fir::EmboxOp> {
+    }
+    // fail: unhandled case
+    TODO();
   }
+};
+
+/// Create a generic box on a memory reference. This conversions lowers the
+/// abstract box to the appropriate, initialized descriptor.
+struct EmboxOpConversion : public EmboxCommonConversion<fir::EmboxOp> {
+  using EmboxCommonConversion::EmboxCommonConversion;
 
   mlir::LogicalResult
   matchAndRewrite(fir::EmboxOp embox, OperandTy operands,
                   mlir::ConversionPatternRewriter &rewriter) const override {
-    auto dims = embox.getDims();
-    assert(!dims);
+    // There should be no dims on this embox op
+    assert(!embox.getDims());
+
     auto loc = embox.getLoc();
     auto *dialect = getDialect();
-    auto ty = unwrap(
-        lowerTy().convertBoxType(embox.getType().dyn_cast<fir::BoxType>(), 0));
+    auto boxTy = embox.getType().dyn_cast<fir::BoxType>();
+    assert(boxTy);
+    auto ty = unwrap(lowerTy().convertBoxType(boxTy, 0));
     auto alloca = genAllocaWithType(loc, ty, defaultAlign, dialect, rewriter);
     auto c0 = genConstantOffset(loc, rewriter, 0);
-    auto rty = unwrap(operands[0].getType()).getPointerTo();
-    auto f0p = genGEP(loc, rty, rewriter, alloca, c0, c0);
-    auto f0p_ = rewriter.create<mlir::LLVM::BitcastOp>(loc, rty, f0p);
-    rewriter.create<mlir::LLVM::StoreOp>(loc, operands[0], f0p_);
-    auto i64Ty = mlir::LLVM::LLVMType::getInt64Ty(dialect);
-    auto i64PtrTy = i64Ty.getPointerTo();
-    auto f1p = genGEPToField(loc, i64PtrTy, rewriter, alloca, c0, 1);
-    auto c0_ = rewriter.create<mlir::LLVM::SExtOp>(loc, i64Ty, c0);
-    rewriter.create<mlir::LLVM::StoreOp>(loc, c0_, f1p);
-    auto i32PtrTy = mlir::LLVM::LLVMType::getInt32Ty(dialect).getPointerTo();
-    auto f2p = genGEPToField(loc, i32PtrTy, rewriter, alloca, c0, 2);
-    rewriter.create<mlir::LLVM::StoreOp>(loc, c0, f2p);
-    auto i8Ty = mlir::LLVM::LLVMType::getInt8Ty(dialect);
-    auto i8PtrTy = mlir::LLVM::LLVMType::getInt8PtrTy(dialect);
-    auto c0__ = rewriter.create<mlir::LLVM::TruncOp>(loc, i8Ty, c0);
-    auto f3p = genGEPToField(loc, i8PtrTy, rewriter, alloca, c0, 3);
-    rewriter.create<mlir::LLVM::StoreOp>(loc, c0__, f3p);
-    auto f4p = genGEPToField(loc, i8PtrTy, rewriter, alloca, c0, 4);
-    rewriter.create<mlir::LLVM::StoreOp>(loc, c0__, f4p);
-    auto f5p = genGEPToField(loc, i8PtrTy, rewriter, alloca, c0, 5);
-    rewriter.create<mlir::LLVM::StoreOp>(loc, c0__, f5p);
-    auto f6p = genGEPToField(loc, i8PtrTy, rewriter, alloca, c0, 6);
-    rewriter.create<mlir::LLVM::StoreOp>(loc, c0__, f6p);
+
+    // Basic pattern to write a field in the descriptor
+    auto storeField = [&](unsigned fldIndex, mlir::Value value,
+                          const std::function<mlir::Value(
+                              mlir::LLVM::LLVMType, mlir::Value)> &applyCast) {
+      auto fldTy = getBoxEleTy(ty, fldIndex);
+      auto fldPtr = genGEPToField(loc, fldTy, rewriter, alloca, c0, fldIndex);
+      auto fld = applyCast(fldTy, value);
```

Comment:
The applyCast callback makes reading the code below a bit harder, would someting like
```
mlir::Value fld;
if (isIntegerType(fldTy)) // or value.getType(), not sure which ways this goes
  fld = integerCast(loc, rewriter, fldTy, value)
else
 fld = rewriter.create<mlir::LLVM::BitcastOp>(loc, fldTy, value)
```
directly in `storeField` works ?

Reply:
I'll see if I can't make this a little more tidy.